### ISSUE TITLE
container with `--pid=host` cause host services be killed

### ISSUE THREAD

**Issue by scuzhanglei** [just-containers/s6-overlay#359]
Dockerfile

```
FROM busybox
ADD https://github.com/just-containers/s6-overlay/releases/download/v1.21.8.0/s6-overlay-amd64.tar.gz /tmp/
RUN gunzip -c /tmp/s6-overlay-amd64.tar.gz | tar -xf - -C /
ENTRYPOINT ["/init"]
```
create a pod with `hostPID=true`，when the pod exit normally would cause services on host be killed，such as sshd。
please be careful when replay this problem on your machine as your sshd service will be killed and you can't connect you machine by ssh anymore。

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: s6
spec:
  hostPID: true
  nodeName: k8s-2
  containers:
  - name: busybox
    args: ["/bin/sh", "-c", "sleep 10"]
    image:  ${docker image build from up Dockerfile}
```

**Comment by skarnet** [comment]
I have no idea what hostPID does, but if it's an option to give access to pids outside of the namespace, then it goes against the very concept of containers - which are, y'know, supposed to _contain_ things. If a process in the container is able to actually send signals and kill processes outside of the container, then it is not a container at all.

And if it is not a container, then I don't understand why anyone would want to use a suite made to handle processes in a container. From my (admittedly incomplete) perspective, the answer is "hostPID should not exist, it's broken by design, and I'm not going to go out of my way to support it". Unless someone can give me a very convincing argument for hostPID, it's a WONTFIX for me.

On a separate note: subreapers are useless, there is no real problem that they solve; and they would not help with this either, reparenting has nothing to do with `kill(-1, sig)` affecting processes outside of the "container".

**Comment by jprjr** [comment]
I'm not super familiar with Kubernetes, but I'm guessing setting that `hostPID` variable is a way to run a Docker container with `--pid=host`.

The use-case I've seen is for running monitoring-type software in a container. One project, `Glances`, has that in their official documentation: https://github.com/nicolargo/glances/blob/e4b4e702e609435092224b049151abf581de11ad/README.rst#docker-the-fun-way

I think that use is better-handled by running the software on the host. At a certain point I think people are just using/abusing Docker to avoid writing service scripts and systemd units for their local installations.

The other use-case is if you wanted to say, run `gdb` or `strace` in a container, but if you're debugging stuff on the host, just run it on the host / use the remote debugging features / etc, and if you need to debug stuff in another container, you can just attach a container to another container's PID namespace (no need to attach to the entire host).

The monitoring-via-container is a pretty niche use-case and in my opinion, better served by installing tools on the actual host. The one thing I can think of is if, say, you're deploying to some abstract Docker-only/Kubernetes-only service? But even then I think they'd provide tools for monitoring what's going on in with your containers.

I guess one other use-case is - you can also attach containers to PID namespaces of other containers. Say your monitoring container attached to the PID namespace of your service container, monitoring container shuts down, performs a `kill(-1,SIGKILL)`, and kills everything in the service container, too. But my answer for that would be, just put your service and monitoring in the same container.

So, I'm fine also marking this as a WONTFIX, I think containers should be as isolated as possible. Attaching to the host's PID namespace really violates that concept. I'm unsure if a container attaching to another's PID namespace also violates that, though?

Re: subreaper, my thought process was (and this is probably getting too complicated) - IF we wanted to support the `--pid=host` case, then instead of `kill(-1,sig)`, the init process could scan for children/grandchildren (maybe by looking at `/proc`?) and send out signals to its own processes. The subreaper would just be a way to ensure they remain child processes.

**Comment by skarnet** [comment]
- About `kill(-1, sig)`: this is something I have _extensively_ thought about, and there is really no simple replacement for it. The only other way to do something similar is to use cgroups, and it's immediately more complicated. Even if you scan `/proc` to seek and destroy all the process trees rooted in your init, you have to do it atomically to avoid race conditions, and you cannot do that without the `freezer` cgroup, which is not supported in cgroupv2. It's a mess; scanning `/proc` is never reliable, and ultimately there is no escaping `kill(-1, sig)`.
- When Linux creates a pid namespace and executes a new pid 1 in it, it automatically makes it a subreaper for all its scions. It's necessary so that this pid 1 actually acts like an init, instead of letting zombies escape to the host. But it's the only case where subreaping is mandatory and the kernel does it for you; there's never a case for _manually_ setting a subreaper with `prctl()`.
- Shall we mark this issue as WONTFIX then, on account of "we assume containers don't leak and you should make sure they don't"?

**Comment by scuzhanglei** [comment]
to make clear, this error is not related to kubernetes, we can reproduce it will docker:
`docker run -it --rm --pid=host`

**Comment by jprjr** [comment]
Well, one final question - in Docker, do we need to use `kill(-1,sig)`? Or is it sufficient to let `s6-svscanctl` handle sending out `SIGTERM`, waiting, then letting Docker clean the rest up?

We may be able to simplify a bit, since once we're in stage3 and sending `SIGTERM`, that's really just trying to gracefully kill any remaining processes that either went rogue, or didn't respond to the original `SIGTERM` in time. We *may* still be capturing their logs with loggers? But I don't think we're going to be running any `finish` scripts, responding to exit codes, etc - it's the very final stages of just trying to shut down cleanly.

If we let Docker handle it instead, we can remove the `s6-nuke` calls, and at the same time support sharing namespaces.

**Comment by skarnet** [comment]
The point is that not all processes are reachable from the supervision tree. Shutting down the supervision tree and all the services under it is easy; but some services may have forked / doubleforked helper processes, or the user init scripts may have spawned an unsupervised long-lived process that we don't know about, etc. The point is, _we don't know_, and short of keeping track via cgroups, the only way to be sure that no process remains alive when we shut down the container is to `kill(-1)` at the end of stage 3.

Even if Docker itself cleans up any remaining process on exit (which, IIRC, it does not, or at least did not in early versions, which is one of the reasons why a real init is needed in containers), it's bad form to leave processes running when shutting down a system. For all we know, they're constantly writing stuff to disk, so we want to try and stop them cleanly before they're brutally SIGKILLed by the kernel when the container manager shuts everything down, in order to avoid filesystem corruption.

On a real machine, you unmount all the filesystems before powering off; you need to kill everything in order to be able to unmount the filesystems. In a container, you cannot unmount all the filesystems or remount them read-only because some of them may be provided by the host and permanently writable, but you _still_ need to kill everything because you still want things to be clean when you shut down.

The alternative is to only shut down the processes that are reachable from the supervision tree, and leave the rest, hoping Docker takes care of them, and hoping they don't cause corruption. That's a lot of hope; and I'm pretty sure there are significantly more cases where these assumptions are false than cases where hostPID is legit.

In short: removing the `kill(-1)` calls has a cost, and I don't think supporting hostPID is worth that cost.

**Comment by skarnet** [comment]
v3.0.0.0 is out! Closing this issue as WONTFIX; please don't use `--pid=host`.
